openfeature/api.py

```diff
@@ -70,18 +71,18 @@ def set_evaluation_context(evaluation_context: EvaluationContext) -> None:
 
 
 def add_hooks(hooks: typing.List[Hook]) -> None:
-    global _hooks
-    _hooks = _hooks + hooks
```

Comment:
Variable assignment should be atomic thanks to the GIL. Not sure I understand the potential race conditions here.

Reply:
that's correct, but you have unpredictable behaviour when multithreaded, because the `_hooks + hooks` creates a complete new list, which has nothing to do with the old one. While locking this part all hooks are added and you can be sure they will all be triggered with any of the hook functions.

Reply:
But isn't that a good thing? Having each version of the list effectively immutable simplifies sharing data safely across threads. With a mutable list I can iterate over the hooks inside the client (after the `_hooks_lock` is released from `get_hooks`) while the list is being modified inside `add_hooks` or `clear_hooks`.

Reply:
I get what you mean, I just feel like my approach will result in the same behaviour as the `providers` code. Both approaches have their benefits.

Reply:
Yeah, I certainly see @federicobond's point here. In code, I guess what he's saying is that some thread is doing something like: 

```python
for hook in get_hooks():
    do_something()
```

while some other thread is doing something like: 

```python
add_hook(MyHook())
```

> Variable assignment should be atomic thanks to the GIL. Not sure I understand the potential race conditions here.

@federicobond I think this stemmed from your comment [here](https://github.com/open-feature/python-sdk/issues/96#issuecomment-1769569433), but I actually think your comment on _this_ PR is correct, that the GIL would prevent these race conditions from occurring. In which case, I think the current implementation is safer than the one implemented in this PR?

Reply:
both approaches should be similar thread safe. Using the lock gives better predictability about the actual behaviour as mentioned. If you add a hook in one thread and get all hooks in another thread, then you will either get all + the one you added or not, depends on how fast Python was able to create the new list and replaced it. Also looking at the current development in Python 3.13 you get the possibility to turn off the GIL 😱 

I talked with @federicobond a while back and we keep the current implementation I'll just remove the unneeded `global` calls in 2 cases.

Reply:
Thanks @gruebel , I have approved the PR.